### PR TITLE
remove ad-hoc 'never' type check in read_lvalue

### LINKED ISSUE #261: Handling of `!`
Currently, miri has a special handling for the `!` type in `read_lvalue` in `lvalues.rs`: The program is aborted whenever a `!` lvalue is read from.

This has me wonder: Why `!`? Is there any special handling in rustc around the `!` type to tell LLVM about unreachable code? Why are we not also, e.g., checking that a `bool` is valid (i.e., `0` or `1` and not anything else) and a reference aligned?

### PR DESCRIPTION
**Pull request by RalfJung**
Funny enough, the correspond test still fails -- thanks to miri checking the storage liveness annotations. ;)

Fixes #261

### CODE REVIEW

File: tests/compile-fail/never_say_never.rs [hunk 1]
```diff
@@ -4,7 +4,7 @@
 fn main() {
     let y = &5;
     let x: ! = unsafe {
-        *(y as *const _ as *const !) //~ ERROR entered unreachable code
+        *(y as *const _ as *const !) //~ ERROR tried to access a dead local variable
```
Comment (eddyb) [inline tests/compile-fail/never_say_never.rs]: I don't understand this error message.

Comment (RalfJung) [inline tests/compile-fail/never_say_never.rs]: This is related to StorageDead and StorageLive. The code seems to mark a variable as StorageDead and then, later, access it. This is, of course, an error.

Comment (RalfJung) [inline tests/compile-fail/never_say_never.rs]: Okay actually what happens is that the variable is just never marked live. We require all variables to be marked live before usage. rustc seemingly never marks variables of type `!` live (which makes sense), leading to this error.

Comment (eddyb) [inline tests/compile-fail/never_say_never.rs]: So... by "access" this means *write*? Also, I am not aware of any conditions for inserting `StorageLive`, this is suspicious.

Comment (RalfJung) [inline tests/compile-fail/never_say_never.rs]: Reads and writes are both accesses, yes. I did not want to introduce separate elements in miri's already way-too-big error type to differentiate read and write...

The rules I implemented for Storage* some weeks ago are that *if* the method uses Storage* at all, then it has to use it for all variables. That's probably stricter than what LLVM requires, but it matches rustc, and it seemed simpler. I think it's quite neat that this manages to catch the `!` here :D

Comment (eddyb) [inline tests/compile-fail/never_say_never.rs]: But I don't see *how* it's possible to catch `!` here, almost seems like an unrelated bug :/.

Comment (RalfJung) [inline tests/compile-fail/never_say_never.rs]: Here's the MIR for main:
```
fn main() -> () {
    let mut _0: ();                      // return pointer
    scope 1 {
        let _1: &i32;                    // "y" in scope 1 at src/main.rs:5:9: 5:10
        scope 2 {
            let _2: !;                   // "x" in scope 2 at src/main.rs:6:9: 6:10
        }
    }
    let mut _3: !;
    let mut _4: *const !;
    let mut _5: *const i32;
    let mut _6: *const i32;
    let mut _7: &i32;
    let mut _8: !;

    bb0: {
        StorageLive(_1);                 // scope 0 at src/main.rs:5:9: 5:10
        _1 = promoted0;                  // scope 0 at src/main.rs:5:13: 5:15
        StorageLive(_2);                 // scope 1 at src/main.rs:6:9: 6:10
        StorageLive(_4);                 // scope 1 at src/main.rs:7:10: 7:37
        StorageLive(_5);                 // scope 1 at src/main.rs:7:11: 7:24
        StorageLive(_6);                 // scope 1 at src/main.rs:7:11: 7:12
        StorageLive(_7);                 // scope 1 at src/main.rs:7:11: 7:12
        _7 = _1;                         // scope 1 at src/main.rs:7:11: 7:12
        _6 = _7 as *const i32 (Misc);    // scope 1 at src/main.rs:7:11: 7:12
        StorageDead(_7);                 // scope 1 at src/main.rs:7:12: 7:12
        _5 = _6;                         // scope 1 at src/main.rs:7:11: 7:24
        StorageDead(_6);                 // scope 1 at src/main.rs:7:24: 7:24
        _4 = _5 as *const ! (Misc);      // scope 1 at src/main.rs:7:10: 7:37
        StorageDead(_5);                 // scope 1 at src/main.rs:7:37: 7:37
        _3 = (*_4);                      // scope 1 at src/main.rs:7:9: 7:37
        _2 = _3;                         // scope 1 at src/main.rs:7:9: 7:37
        StorageDead(_3);                 // scope 1 at src/main.rs:7:37: 7:37
        StorageDead(_4);                 // scope 1 at src/main.rs:8:7: 8:7
        _8 = _2;                         // scope 2 at src/main.rs:9:7: 9:8
        const f(_8);                     // scope 2 at src/main.rs:9:5: 9:9
    }
}
```
Notice that _3 is never marked live. It hence remains dead, which makes the write to it invalid.

Comment (eddyb) [inline tests/compile-fail/never_say_never.rs]: There's a `StorageDead` for it. That seems like a bug in the MIR generation.

Comment (eddyb) [inline tests/compile-fail/never_say_never.rs]: See https://github.com/rust-lang/rust/commit/84d1f6aa82123c2951042aff99e43bdb894d3d81 - this is the wrong microoptimizations IMO, one could just emit `unreachable` instead of creating the temporary at all.

Comment (oli-obk) [inline tests/compile-fail/never_say_never.rs]: It should also emit unreachable for all uninhabited types.

Comment (RalfJung) [inline tests/compile-fail/never_say_never.rs]: That's actually still being discussed. There are some subtle issues here... see <https://internals.rust-lang.org/t/recent-change-to-make-exhaustiveness-and-uninhabited-types-play-nicer-together/4602/75> for example.

**Comment by oli-obk** [comment]
So... @eddyb should we merge this and wait for rustc to fix it?
